src/metaswitch/clearwater/queue_manager/queue_fsm.py

```diff
@@ -0,0 +1,138 @@
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or (at your
+# option) any later version, along with the "Special Exception" for use of
+# the program along with SSL, set forth below. This program is distributed
+# in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+# without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU General Public License for more
+# details. You should have received a copy of the GNU General Public
+# License along with this program.  If not, see
+# <http://www.gnu.org/licenses/>.
+#
+# The author can be reached by email at clearwater@metaswitch.com or by
+# post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+#
+# Special Exception
+# Metaswitch Networks Ltd  grants you permission to copy, modify,
+# propagate, and distribute a work formed by combining OpenSSL with The
+# Software, or a work derivative of such a combination, even if such
+# copying, modification, propagation, or distribution would otherwise
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+import constants
+from .queue_state import QueueStateInfo
+from .alarms import QueueAlarm
+from .timers import QueueTimer
+import logging
+from functools import partial
+from queue_utils import move_to_processing, mark_node_as_errored
+
+_log = logging.getLogger("queue_manager.queue_fsm")
+
+class QueueFSM(object):
+
+    def __init__(self, plugin, node_id, callback_func):
+        self._queue_states = None
+        self._queue_config = {}
+        self._last_local_state = None
+
+        self._plugin = plugin
+        self._id = node_id
+        self._running = True
+
+        self._timer_callback_func = callback_func
+        self._timer = None
+        self._local_alarm = QueueAlarm(*self._plugin.local_alarm())
+        self._global_alarm = QueueAlarm(*self._plugin.global_alarm())
+
+        self._local_fsm = {constants.LS_NO_QUEUE: [self._local_alarm.clear],
+                           constants.LS_NO_QUEUE_ERROR: [self._local_alarm.critical],
+                           constants.LS_FIRST_IN_QUEUE: [self._local_alarm.minor,
+                                                         partial(move_to_processing, self.return_queue_config, self._id)],
+                           constants.LS_PROCESSING: [self._plugin.at_front_of_queue,
```

Comment:
Worth a comment to say that these functions are all called, even if one of them changes the local state?